src/progress/inflights.rs

```diff
@@ -0,0 +1,267 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Copyright 2015 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// A buffer of inflight messages.
+#[derive(Debug, PartialEq)]
+pub struct Inflights {
+    // the starting index in the buffer
+    start: usize,
+    // number of inflights in the buffer
+    count: usize,
+
+    // ring buffer
+    buffer: Vec<u64>,
+}
+
+// The `buffer` must have it's capacity set correctly on clone, normally it does not.
+impl Clone for Inflights {
+    fn clone(&self) -> Self {
+        let mut buffer = self.buffer.clone();
+        buffer.reserve(self.buffer.capacity() - self.buffer.len());
+        Inflights {
+            start: self.start,
+            count: self.count,
+            buffer,
+        }
+    }
+}
+
+impl Inflights {
+    /// Creates a new buffer for inflight messages.
+    pub fn new(cap: usize) -> Inflights {
+        Inflights {
+            buffer: Vec::with_capacity(cap),
+            start: 0,
+            count: 0,
+        }
+    }
+
+    /// Returns true if the inflights is full.
+    pub fn full(&self) -> bool {
```

Comment:
seem we can use inline for all simple functions.

Reply:
That's right. I'll do the checking.